Internal refactor, introduce `AtRule`

RobinMalfait opened the pull request:
This PR introduces an internal refactor where we introduce the `AtRule` CSS Node in our AST.

The motivation for this is that in a lot of places we need to differentiate between a `Rule` and an `AtRule`. We often do this with code that looks like this:

```ts
rule.selector[0] === '@' && rule.selector.startsWith('@media')
```

Another issue we have is that we often need to check for `'@media '` including the space, because we don't want to match `@mediafoobar` if somebody has this in their CSS. Alternatively, if you CSS is minified then it could be that you have a rule that looks like `@media(width>=100px)`, in this case we _also_ have to check for `@media(`.

Here is a snippet of code that we have in our codebase:

```ts
// Find at-rules rules
if (node.kind === 'rule') {
  if (
    node.selector[0] === '@' &&
    (node.selector.startsWith('@media ') ||
      node.selector.startsWith('@media(') ||
      node.selector.startsWith('@custom-media ') ||
      node.selector.startsWith('@custom-media(') ||
      node.selector.startsWith('@container ') ||
      node.selector.startsWith('@container(') ||
      node.selector.startsWith('@supports ') ||
      node.selector.startsWith('@supports(')) &&
    node.selector.includes(THEME_FUNCTION_INVOCATION)
  ) {
    node.selector = substituteFunctionsInValue(node.selector, resolveThemeValue)
  }
}
```

Which will now be replaced with a much simpler version:
```ts
// Find at-rules rules
if (node.kind === 'at-rule') {
  if (
    (node.name === '@media' ||
      node.name === '@custom-media' ||
      node.name === '@container' ||
      node.name === '@supports') &&
    node.params.includes(THEME_FUNCTION_INVOCATION)
  ) {
    node.params = substituteFunctionsInValue(node.params, resolveThemeValue)
  }
}
```

Checking for all the cases from the first snippet is not the end of the world, but it is error prone. It's easy to miss a case.

A direct comparison is also faster than comparing via the `startsWith(…)` function.

---

Note: this is only a refactor without changing other code _unless_ it was required to make the tests pass. The tests themselves are all passing and none of them changed (because the behavior should be the same).
The one exception is the tests where we check the parsed AST, which now includes `at-rule` nodes instead of `rule` nodes when we have an at-rule.

RobinMalfait committed:
introduce `AtRule`

thecrypticace commented on packages/tailwindcss/src/index.ts:
```diff
@@ -381,15 +380,18 @@ async function parseCss(
         // Wrap `@keyframes` in `AtRoot` so they are hoisted out of `:root` when
         // printing.
         nodes.push(
-          Object.assign(keyframesRule, atRoot([rule(keyframesRule.selector, keyframesRule.nodes)])),
+          Object.assign(
+            keyframesRule,
+            atRoot([atRule(keyframesRule.name, keyframesRule.params, keyframesRule.nodes)]),
+          ),
         )
```
Do you know if this can be `nodes.push(atRoot([keyframesRule]))`?

RobinMalfait replied:
Just ported it as-is, but if we just use `Object.assign(keyframesRule, atRoot([keyframesRule]))` then we get an infinite loop (maybe that's why we did it this way).

That said, the `nodes.push(…)` you mentioned seems to work as expected.

RobinMalfait committed:
rename `Rule` to `StyleRule`

This follows the CSS spec naming, which is useful if we have `StyleRule`
and `AtRule`.

RobinMalfait committed:
add `rule(…)` ast helper

This keeps the codebase in a similar state as before, but this just
creates one of the follow nodes: `AtRule | StyleRule`

RobinMalfait committed:
run prettier

RobinMalfait committed:
keep the `'rule'` name

Keeps the diff smaller. Might drop this commit before merging.

RobinMalfait committed:
use `rule(…)` in more spots

This keeps the diff smaller

RobinMalfait committed:
use `styleRule` if we _know_ it's a style rule

We can use `rule(…)` here as well, and it will make the diff smaller.
However, that requires us to jump through an additional function where
we verify that we don't start with `@`.

We already know this, so let's skip the indirection.

RobinMalfait commented:
@adamwathan haha that was my next thing on the list to try. I'm hoping for the same results you mentioned.

RobinMalfait commented:
Running some benchmarks on the Tailwind UI codebase.

Before (`next`):
```
 ✓ src/index.bench.ts (1) 667ms
     name          hz      min      max     mean      p75      p99     p995     p999     rme  samples
   · compile  54.5384  16.6713  22.1950  18.3357  18.7034  22.1950  22.1950  22.1950  ±2.45%       28
```

After (`pr`):
```
 ✓ src/index.bench.ts (1) 650ms
     name          hz      min      max     mean      p75      p99     p995     p999     rme  samples
   · compile  53.4548  16.5493  28.0242  18.7074  18.9284  28.0242  28.0242  28.0242  ±5.57%       27
```

It looks like this PR is a tiny bit slower right now. Thinking about it, it makes sense from a CSS parsing perspective because we are doing a bit more when we see at-rules. We also re-parse every time we see an at-rule in the variants used, I think this is where we can do some speedups. 

Let me try some things, because it's easier to work with style rules and at rules separately, but not at the cost of perfomance.

RobinMalfait committed:
replace regex search with good ol' for loop

RobinMalfait committed:
always skip 6 characters

This is a tiny optimization, but the moment we call `parseAtRule` we
know 2 things:
1. It starts with an `@` because it's an at-rule, therefore we can skip
   this character.
2. You need at least 4 characters (excluding `@`) because `@page` is the
   shortest valid at-rule.

This means that we can start looking past `@page`, which means that we
can start at character 6.

List of valid CSS at-rules (currently) ordered by length:

- `@page`
- `@layer`
- `@media`
- `@scope`
- `@import`
- `@charset`
- `@property`
- `@supports`
- `@container`
- `@font-face`
- `@keyframes`
- `@namespace`
- `@custom-media`
- `@position-try`
- `@color-profile`
- `@counter-style`
- `@starting-style`
- `@view-transition`
- `@font-feature-values`
- `@font-palette-values`

RobinMalfait committed:
use `charCodeAt` instead of string compare

RobinMalfait committed:
push the wrapped keyframe directly

No need to create a new rule and mutate the original. We can just push
the new rule directly and share the original keyframesRule object.

RobinMalfait commented:
Few improvements in, these are the results:

Before (`next`):
```
 ✓ src/index.bench.ts (1) 666ms
     name          hz      min      max     mean      p75      p99     p995     p999     rme  samples
   · compile  52.8106  16.1739  29.1533  18.9356  18.9055  29.1533  29.1533  29.1533  ±6.61%       27
```

After (`pr`):
```
 ✓ src/index.bench.ts (1) 661ms
     name          hz      min      max     mean      p75      p99     p995     p999     rme  samples
   · compile  52.8260  16.2722  29.1343  18.9301  19.7049  29.1343  29.1343  29.1343  ±6.34%       27
```

RobinMalfait committed:
let's only skip 2 characters, e.g.: `@x`

In theory, people could write their own at-rules (e.g.: if you sandwich
`@tailwindcss/postcss` between 2 PostCSS plugins that handle custom
at-rules).

In that scenario, it is possible to have `@x` at-rules.

RobinMalfait committed:
skip 5 characters

Changed my mind again. There are a lot of conditions necessary to use
shorter at-rules. This means that we can improve the parsing by using
the more common case of skipping 5 characters by default.

We can adjust this number if it's actually causing issues.

adamwathan commented on packages/tailwindcss/src/ast.ts:
```diff
@@ -1,9 +1,20 @@
+
+const AT_SIGN = 0x40
+
+export type StyleRule = {
```
I think we should just use `Rule` for this type and `rule` for the helper function instead of `styleRule`, as rules and at-rules are distinct concepts in CSS terminology already:

<img width="853" alt="image" src="https://github.com/user-attachments/assets/83d93f9f-dec6-40f0-baee-5a48c9493fec">

RobinMalfait replied:
I was thinking the same, but @thecrypticace pointed me to the spec where they use Style rules and At rules. I think _just_ Rule is very common because of PostCSS.

- https://www.w3.org/TR/css-syntax-3/#css-stylesheets
 
<img width="246" alt="image" src="https://github.com/user-attachments/assets/61777034-1bb1-47eb-af69-256e7375b1a5">

I'm fine with either. Biggest benefit for separating is that we have a `rule()` abstraction that returns a style rule via `styleRule()` or an at rule via `atRule()` right now depending on the `@` being present.

thecrypticace replied:
Yeah the spec *explicitly* calls them style rules.

adamwathan replied:
Yeah I'm fine with style rule then if that's considered more technically correct 👍

adamwathan reviewed (commented):
Nice! I think this is a great improvement.

The only high level thing I'm half wondering about is whether we should include the `@` sign as part of `AtRule.name` for greppability even if it's not necessary? Don't have a strong opinion on it but does feel like something we're losing with this change. Could easily be convinced not to worry about it.

RobinMalfait commented:
Hmm, I do like the grepability of looking for `@media` 🤔

Fun fact, during printing, I forgot to add an `@` in some spots, so maybe this is not a bad change.

Any strong feelings about this @thecrypticace?

RobinMalfait committed:
keep `@` in at-rule name

This allows us to just search for `@media` in the codebase and find
wherever we used the at-rules. This is purely a development QoL
improvement.

adamwathan reviewed (approved):
Glad we decided to include the `@` seeing the diff now, I think the code reads a lot better with it being part of the string 👍
